### PR TITLE
[KVStore] New version of Distributed kvstore

### PR DESCRIPTION
**Pull request by aksnzhy**
## Description

This PR is the new version of distributed kvstore based on our new network communicator, which is more efficient than the original one.

### Initialization

DGL kvstore contains two components: kvserver and kvclient. We can create kvserver and kvclient by using the following APIs:

IP namebook:

```python
client_namebook = { 0:'127.0.0.1:50051',
                    1:'127.0.0.1:50052' }

server_namebook = { 0:'127.0.0.1:50053',
                    1:'127.0.0.1:50054' }
```

We can read name book by `ReadNetworkCongfigure()` API.

kvserver:

```python
server = dgl.contrib.KVServer(
        server_id=args.id, 
        client_namebook=client_namebook, 
        server_addr=server_namebook[args.id])

server.start()
```

kvclient:

```python
client = dgl.contrib.KVClient(
        client_id=args.id, 
        server_namebook=server_namebook, 
        client_addr=client_namebook[args.id])

client.connect()
```

We can start multiple kvserver and kvclient on different machines, and each server/client will have its own unique ID.

 The basic data structure transferred between kvserver and kvclient is tensor (mx.ndarray or torch.tensor). We can initialize tensor of server by client side  using the following API:

```python
client.init_data(name='embed', shape=[10, 3], init_type='uniform', low=-1.0, high=1.0)
```
This API initialize kvstore tensor with random values in uniform distribution.

We can also initialize all data to zero:

```python
client.init_data(name='embed', shape=[10, 3], init_type='zero)
```
### PUSH / PULL

After initialization, kvclient can push/pull data to/from kvserver. 

For example:

```python
tensor_id = torch.tensor([0, 1, 2])
tensor_data = torch.tensor([[0., 0., 0., ], [1., 1., 1.], [2., 2., 2.]])
client.push('embed', tensor_id, tensor_data)
new_tensor = client.pull('embed', tensor_id)
```

### Barrier

The client can invoke `barrier()` API for synchronization. 

For example:

```python
tensor_id = torch.tensor([0, 1, 2])
tensor_data = torch.tensor([[0., 0., 0., ], [1., 1., 1.], [2., 2., 2.]])
client.push('embed', tensor_id, tensor_data)
client.barrier()
new_tensor = client.pull('embed', tensor_id)
```

### Computation on KVStore

On default, the kvserver will do `add` operation when receiving a push message, and it will do `gather_row` operation when receiving a pull message. Developers can re-implement these two default operations by override _push_handler and _pull_handler on kvserver. 

```python
class MyKVServer(KVServer):
    def _push_handler(self, name, ID, data):
          #... User-defined code ... 
    def _pull_handler(self, name, ID):
          #... User-defined code ... 
```

### Data Partition Scheme

DGL kvstore uses a very simple range-partition scheme to partition data into different server nodes. For example, if the total embedding count is 200 and we have two kvserver, the data [0-99] will be stored in kvserver_0, and data [100-199] will be stored in kvserver_1.  The push() and pull() API will mapping data to corresponded server node automatically.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/mxnet/dis_kvstore/client.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Initialize data on server
+    client.init_data(name='embed_0', shape=[10, 3], init_type='zero')
+    client.init_data(name='embed_1', shape=[11, 3], init_type='uniform', low=0.0, high=0.0)
+
+    tensor_id = mx.nd.array([0, 1, 2], dtype='int64')
+    tensor_data = mx.nd.array([[0., 0., 0., ], [1., 1., 1.], [2., 2., 2.]])
+
+    # Push data
+    for i in range(5):
```
Comment (classicsong) [inline examples/mxnet/dis_kvstore/client.py]: Does this for loop has any meaning?

Comment (aksnzhy) [reply examples/mxnet/dis_kvstore/client.py]: > Does this for loop has any meaning?

No, I just want to test sending data 5 times.

File: python/dgl/contrib/dis_kvstore.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        ----------
+        name : str
+            data name
+        ID : tensor (mx.ndarray or torch.tensor)
+            a vector storing the global IDs
+        data : tensor (mx.ndarray or torch.tensor)
+            a matrix with the same row size of id
+        """
+        assert F.ndim(ID) == 1, 'ID must be a vector.'
+        assert F.ndim(data) == 2, 'data must be a matrix.'
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py]: i think 3D, 4D or even higher is fine.

File: python/dgl/contrib/dis_kvstore.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+        ------
+        tensor
+            a matrix with the same row size of ID
+        """
+        assert F.ndim(ID) == 1, 'ID must be a vector.'
+        group_size = [0] * self._server_count
+        numpy_id = F.asnumpy(ID)
+        for id in numpy_id:
+            server_id = self._get_server_id(id, name)
+            group_size[server_id] += 1
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py]: is there a better way of computing this? it can be slow.

File: python/dgl/contrib/dis_kvstore.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        min_idx = 0
+        max_idx = 0
+        server_count = 0
+        for idx in range(self._server_count):
+            if group_size[idx] == 0:
+                continue
+            server_count += 1
+            max_idx += group_size[idx]
+            range_id = ID[min_idx:max_idx]
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py]: You assume the row Ids in `ID` is in the ascending order. Is it a right assumption?

File: python/dgl/contrib/dis_kvstore.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+            msg = _recv_kv_msg(self._receiver)
+            assert msg.type == KVMsgType.PULL_BACK, 'Recv kv msg error.'
+            msg_list.append(msg)
+
+        return self._merge_msg(msg_list)
+    
+    def barrier(self):
+        """Barrier for all client nodes
+        """
+        pass
```
Comment (zheng-da) [inline python/dgl/contrib/dis_kvstore.py]: is it not implemented yet?

File: python/dgl/network.py [hunk 6]
```diff
@@ -1,6 +1,12 @@
 """DGL Distributed Training Infrastructure."""
 from __future__ import absolute_import
 
+import time
```
Comment (zheng-da) [inline python/dgl/network.py]: up to here.

File: python/dgl/contrib/launch_schedule/dgl_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+#
+# 1. Copy the executable files, data files and some scripts to a temporary directory
+#    in each remote machine, the temporary directory is specified by user.
+#
+# 2. Start communictors in each machine by running python script 'worker.py',
+#    which will establish communication among machines
+#
+# 3. Start server-worker or client-worker for each machine.
+
+# To be done
```
Comment (zheng-da) [inline python/dgl/contrib/launch_schedule/dgl_scheduler.py]: i suppose these are the scripts that launch kvstore and model training in a cluster? maybe we can remove them from this PR to make this PR smaller.

File: python/dgl/network.py [hunk 8]
```diff
@@ -153,3 +174,121 @@ def _recv_nodeflow(receiver, graph):
...
+
+
+class KVMsgType(Enum):
+    """Type of kvstore message
+    """
+    FINAL = 1
+    INIT = 2
+    PUSH = 3
+    PULL = 4
+    PULL_BACK = 5
```
Comment (zheng-da) [inline python/dgl/network.py]: what is pull back?

File: python/dgl/network.py [hunk 9]
```diff
@@ -153,3 +174,121 @@ def _recv_nodeflow(receiver, graph):
...
+            msg.name,
+            tensor_id)
+    elif msg.type in (KVMsgType.FINAL, KVMsgType.BARRIER):
+        _CAPI_SenderSendKVMsg(
+            sender,
+            int(recv_id),
+            msg.type.value,
+            msg.rank)
+    else:
+        tensor_id = F.zerocopy_to_dgl_ndarray(msg.id)
```
Comment (zheng-da) [inline python/dgl/network.py]: So INIT, PUSH, PULL_BACK are handled here?

File: src/graph/network.cc [hunk 10]
```diff
@@ -24,7 +24,32 @@ using namespace dgl::runtime;
...
+    tensor.strides[i] = 1;
+  }
+  for (int i = tensor.ndim - 2; i >= 0; --i) {
+    tensor.strides[i] = tensor.shape[i+1] * tensor.strides[i+1];
+  }
+  tensor.data = raw;
+  DLManagedTensor *managed_tensor = new DLManagedTensor();
+  managed_tensor->dl_tensor = tensor;
+  return NDArray::FromDLPack(managed_tensor);
+}
```
Comment (zheng-da) [inline src/graph/network.cc]: No NDArray constructor creates an array in this way?

Comment (zheng-da) [reply src/graph/network.cc]: I suppose the created array doesn't own the memory. @jermainewang is this a right way?

File: src/graph/network.cc [hunk 11]
```diff
@@ -88,6 +113,54 @@ void MsgMeta::Deserialize(char* buffer, int64_t size) {
...
+char* KVStoreMsg::Serialize(int64_t* size) {
+  char* buffer = nullptr;
+  int64_t buffer_size = 0;
+  buffer_size += sizeof(this->msg_type);
+  buffer_size += sizeof(this->rank);
+  if (!this->name.empty()) {
+    buffer_size += sizeof(this->name.size());
+    buffer_size += this->name.size();
+  }
+  buffer = new char[buffer_size];
```
Comment (zheng-da) [inline src/graph/network.cc]: Can you put a comment here? In the future, we should have a better memory management. allocating a large chunk of memory can be very expensive.

File: src/graph/network.h [hunk 12]
```diff
@@ -24,38 +25,66 @@ namespace network {
...
+   */
   kNodeFlowMsg = 0,
- /*!
-  * \brief Message for end-signal
-  */
-  kEndMsg = 1
+  /*!
+   * \brief Message for end-signal
+   */
+  kEndMsg = 1,
```
Comment (zheng-da) [inline src/graph/network.h]: please make the name consistent with the one in Python. In Python, it's called Final

File: src/graph/network.cc [hunk 13]
```diff
@@ -194,182 +267,312 @@ DGL_REGISTER_GLOBAL("network._CAPI_SenderSendNodeFlow")
...
+        kv_msg.data = args[6];
+      }
+    }
+    int64_t kv_size = 0;
+    char* kv_data = kv_msg.Serialize(&kv_size);
+    // Send kv_data
+    Message send_kv_msg;
+    send_kv_msg.data = kv_data;
+    send_kv_msg.size = kv_size;
+    send_kv_msg.deallocator = DefaultMessageDeleter;
```
Comment (zheng-da) [inline src/graph/network.cc]: I don't understand. Here it seems you send data. What is the data message below. Please document every message here. Why are there so many messages?

File: examples/mxnet/dis_kvstore/client.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        client.push('embed_0', tensor_id, tensor_data)
+        client.push('embed_1', tensor_id, tensor_data)
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply examples/mxnet/dis_kvstore/client.py]: Yes, it is sparse row ids

File: examples/mxnet/dis_kvstore/client.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+    tensor_id = mx.nd.array([6, 7, 8])
+    for i in range(5):
+        client.push('embed_0', tensor_id, tensor_data)
+        client.push('embed_1', tensor_id, tensor_data)
+
+    time.sleep(1) # wait all Push done
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply examples/mxnet/dis_kvstore/client.py]: We can use `client.barrier()` here.

File: python/dgl/contrib/dis_kvstore.py [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+        """
+        _finalize_sender(self._sender)
+        _finalize_receiver(self._receiver)
+
+    def start(self):
+        """Start service of KVServer
+        """
+        server_ip, server_port = self._addr.split(':')
+        _receiver_wait(self._receiver, server_ip, int(server_port), self._client_count)
+        _network_wait() # wait client's start
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py]: It is used to make sure that when server connect back to client, everything is done on client (typically 1s is enough).

File: python/dgl/contrib/dis_kvstore.py [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+        _finalize_sender(self._sender)
+        _finalize_receiver(self._receiver)
+
+    def connect(self):
+        """Connect to all KVServer nodes
+        """
+        for ID, addr in self._server_namebook.items():
+            server_ip, server_port = addr.split(':')
+            _add_receiver_addr(self._sender, server_ip, int(server_port), ID)
+        _sender_connect(self._sender)
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py]: If the server is started before the client, there is no problem. If the server is not started before the client, the client will re-try the connection in a loop until the server is started.

File: python/dgl/network.py (hunk 8 above)
Comment (aksnzhy) [reply python/dgl/network.py]: PULL_BACK is the message sent from server when the client invoke a PULL message.

File: python/dgl/contrib/dis_kvstore.py [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+        # Service loop
+        while True:
+            msg = _recv_kv_msg(self._receiver)
+            if msg.type == KVMsgType.INIT:
+                if (msg.name in self._is_init) == False:
+                    # we hack the msg format here
+                    data_shape = F.asnumpy(msg.id).tolist()
+                    array_0 = (F.asnumpy(msg.data).tolist())[0] 
+                    array_1 = (F.asnumpy(msg.data).tolist())[1]
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py]: I write some comments here.

File: python/dgl/contrib/dis_kvstore.py [hunk 19]
```diff
@@ -0,0 +1,487 @@
...
+                    init_type = 'zero' if array_0[0] == 0.0 else 'uniform'
```
[the remark this answers is not in the document]
Comment (aksnzhy) [reply python/dgl/contrib/dis_kvstore.py]: I hope we can maintain a minimal data structure for kvstore message, so I didn't add two many data fields into message structure.

File: python/dgl/network.py (hunk 9 above)
Comment (aksnzhy) [reply python/dgl/network.py]: Here is just sending different type of message.

Comment (aksnzhy) [reply python/dgl/network.py]: The real handler is in the while Loop on service.

File: src/graph/network.cc (hunk 13 above)
Comment (aksnzhy) [reply src/graph/network.cc]: We have two types of messages here. `KVStoreMsg` is a wrapper for python message. `Message` is used for communicator, which is a pure binary message.

**Comment by aksnzhy** [comment]
@zheng-da There is something wrong with this PR and I re-open a PR.

**Pull request closed without merging**